### PR TITLE
fill out the visitor for BoundMethodGroup

### PR DESCRIPTION
**Pull request by TyOverby**
Closes https://github.com/dotnet/roslyn/issues/13734

### CODE REVIEW

**Comment by TyOverby** [comment]
@dotnet/roslyn-compiler @gafter 

This fixes a compiler crash on expressions that have the AST `Conversion(MethodGroup(Await(...)))`

```cs
Func<Type> f = (await Task.FromResult(1)).GetType;
```

**Comment by jaredpar** [comment]
Which bug is this addressing?

File: src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs [hunk 1]
```diff
@@ -3857,5 +3857,28 @@ static void M()
...
+        public void TypeofAwaitedTask()
+        {
+            string source =
+    @"
+using System.Threading.Tasks;
+using System;
+
+namespace AsyncBug
+{
+    public class Foo {
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs]: Please avoid `Foo` in bug repros.  It's silly but it's a policheck violation that we have to excuse later on.  Easier to just avoid it.

Says the guy who had to run poli-check on all our source code before moving to GitHub ancd dealing with excusing the couple thousand references to `Foo`, `Bar` and `FooBar` that we had in the code base.

File: src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs [hunk 2]
```diff
@@ -3857,5 +3857,28 @@ static void M()
...
+        public async void Boom() {
+            Func<Type> f = (await Task.FromResult(1)).GetType;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs]: We should emit IL for this test to verify that we properly spill the temporary here.

**Comment by TyOverby** [comment]
@jaredpar The PR description has been updated to list [the issue](https://github.com/dotnet/roslyn/issues/13734)

File: src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs (hunk 1 above)
Comment (TyOverby) [inline src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs]: Done.

(there's 124 other instances of Foo in this file)

File: src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs (hunk 2 above)
Comment (TyOverby) [inline src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs]: Now that I think about it, this test case should probably be grouped with the other async emit tests.

**Comment by TyOverby** [comment]
@jaredpar: @VSadov and I came up with two behavioral tests that validate the spilling behavior.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncSpillTests.cs [hunk 3]
```diff
@@ -2934,5 +2934,75 @@ public static void Main()
...
+        private class SomeClass
+        {
+            public bool Method(int value)
+            {
+                return value % 2 == 0;
+            }
+        }
+
+        private Task<SomeClass> Danger()
+        {
```
Comment (VSadov) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncSpillTests.cs]: could you make this method async and add ```await Task.Yield()``` as a first statement - that is to make sure that callee gets suspended and to make sure spilling/re-hydrating is involved.

With just ```FromResult```  Task, it is very likely that the call returns synchronously.

**Comment by AlekseyTs** [comment]
LGTM, modulo the test suggestion from @VSadov.

**Comment by AlekseyTs** [comment]
Please test similar scenario for VB.

**Comment by TyOverby** [comment]
@AlekseyTs @VSadov 

The latest commit has fixes for your suggestions in them.

**Comment by VSadov** [comment]
LGTM

**Comment by AlekseyTs** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AwaitExpressionSpiller.cs [hunk 4]
```diff
@@ -854,6 +854,13 @@ public override BoundNode VisitConversion(BoundConversion node)
                     type: node.Type));
         }
 
+        public override BoundNode VisitMethodGroup(BoundMethodGroup node)
+        {
+            BoundSpillSequenceBuilder builder = null;
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AwaitExpressionSpiller.cs]: >BoundSpillSequenceBuilder [](start = 12, length = 25)

LGTM

This fix is correct, but a method group should not get this far. This has exposed a flaw in lowering, which should turn the method group conversion into a delegate creation. I'll open a (design) issue.

**Comment by jasonmalinowski** [comment]
@dotnet-bot test windows_release_unit32_prtest please and test windows_release_unit64_prtest please. (These are new jobs that are now also required.)

**Pull request merged**